### PR TITLE
Fix assert; arguments were in reverse order

### PR DESCRIPTION
**Pull request by Connor-GH**
So sometime between glibc 2.35 and glibc 2.38, assert for betterC stopped working.
After a wild goosechase, it was determined that not only was the wrong function being
called ("__assert" instead of "__assert_fail"), but the arguments were being passed to
el_bin backwards! The fix for this is obviously to fix the order of the arguments, and
this is reflected in the generated code. Musl was untested for this, but the fix should
not affect otherwise working musl systems due to ``version(CRuntime_Musl)``.

### CODE REVIEW

**Comment by Connor-GH** [comment]
Whoops, I accidently closed this.

**Comment by schveiguy** [comment]
This needs a test. And I suspect it might be dependent on the glibc version.

In discord chat, we found that on my glibc 2.35 version, it worked, but on @Connor-GH 's glibc 2.38 version it had a segfault.

If this is the case, it's going to be a tricky thing to get right.

File: compiler/src/dmd/e2ir.d
Comment (ibuclaw) [inline compiler/src/dmd/e2ir.d:7031]:
```suggestion
        else version (CRuntime_Glibc)
```

File: compiler/src/dmd/e2ir.d
Comment (ibuclaw) [inline compiler/src/dmd/e2ir.d:7046]:
```suggestion
            // [_]_assert(msg, file, line);
            const rtlsym = (irs.target.os == Target.OS.Windows) ? RTLSYM.C_ASSERT : RTLSYM.C__ASSERT;
            auto eassert = el_var(getRtlsym(rtlsym));
            ea = el_bin(OPcall, TYvoid, eassert, el_params(eline, efilename, elmsg, null));
```

File: compiler/src/dmd/e2ir.d
Comment (ibuclaw) [inline compiler/src/dmd/e2ir.d:7036]: So are params passed left-to-right or right-to-left?

This invocation differs from musl, so both can't be right.

**Comment by ibuclaw** [comment]
> This needs a test. And I suspect it might be dependent on the glibc version.
> 

Assuming that there aren't any per-arch overrides, implementation hasn't changed since 2002.

https://github.com/bminor/glibc/blob/f83e461f1014598a5cb4c89407ce303b9f0bd8ac/assert/__assert.c#L23

> In discord chat, we found that on my glibc 2.35 version, it worked, but on @Connor-GH 's glibc 2.38 version it had a segfault.
> 
> If this is the case, it's going to be a tricky thing to get right.
> 
> Edit: Still needs a test, but I'm realizing this is a different function being called! It's `__assert_fail` and not `__assert`, which we agreed is the correct call.

Arguably both are wrong because this is relying on glibc internals that are not intended to be called directly like this.

**Review by thewilsonator — CHANGES_REQUESTED** [review]
we should be using `irs.target.c.runtime` here not `version(Runtime_Foo)`

```d
if (irs.target.os == Target.OS.OSX) { ... }
else
{
    Symbol* assertSym;
    elem* params;
    with (TargetC.Runtime) switch  (irs.target.c.runtime)
    {
        case Musl:
            assertSym = getRtlsym(RTLSYM.C__ASSERT_FAIL);
            params = el_params(elmsg, eline, efilename, efunc, null);
            break;
        case GLibc: 
            //...
    }
    elem* efunc = getFuncName();
    auto eassert = el_var(assertSym);
    ea = el_bin(OPcall, TYvoid, eassert, params);
}
```

File: compiler/src/dmd/e2ir.d
Comment (Connor-GH) [reply compiler/src/dmd/e2ir.d:7031]: I tried this but serve-d was showing syntax errors, so I didn't try to compile it. This is what I meant to do.

**Comment by Connor-GH** [comment]
I'd also like to mention that the CI is pretty outdated. None of the CI machines run glibc 2.38.

**Comment by Connor-GH** [comment]
Tested on void linux musl, the assert was backwards for musl too. That means it might also be backwards for OSX.

**Comment by schveiguy** [comment]
> Arguably both are wrong because this is relying on glibc internals that are not intended to be called directly like this.

Agreed. It is something I was thinking importC might actually be better at solving. But of course, this then *requires* a C compiler for building anything.

File: compiler/src/dmd/e2ir.d
Comment (ibuclaw) [inline compiler/src/dmd/e2ir.d:7043]:
```suggestion
        auto eassert = el_var(assertSym);
        ea = el_bin(OPcall, TYvoid, eassert, params);
```

Looks like just a small typo.

**Review by Connor-GH — COMMENTED** [review]
I have implemented this.

File: compiler/src/dmd/e2ir.d
Comment (Connor-GH) [reply compiler/src/dmd/e2ir.d:7043]: Good call! I did not mean to leave that in there.

**Comment by Connor-GH** [comment]
Simply waiting on approval now for CI.

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Is there a bugzilla issue number for this? if not please create one, or add a changelog entry for this. This otherwise looks good, the one failure I think will go away if you rebase this PR

**Comment by Connor-GH** [comment]
https://issues.dlang.org/show_bug.cgi?id=24558

**Pull request merged**
